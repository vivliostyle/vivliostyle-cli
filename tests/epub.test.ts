// Mocked modules: don't reorder this list
import './mocks/fs.js';
import './mocks/tmp.js';
//
import './mocks/archiver.js';

import AdmZip from 'adm-zip';
import { vol } from 'memfs';
import { format } from 'prettier';
import { assert, beforeEach, expect, it } from 'vitest';
import { exportEpub } from '../src/output/epub.js';
import { decodePublicationManifest } from '../src/output/webbook.js';
import { PublicationManifest } from '../src/schema/publication.schema.js';
import { runCommand, toTree } from './command-util.js';

beforeEach(() => {
  vol.reset();
});

function checkValidEpubZip(epub: Buffer) {
  // Check epub file contains uncompressed mimetype file
  expect(epub.readUInt32BE(0)).toBe(0x504b0304);
  expect(epub.readUInt16LE(8)).toBe(0);
  expect(epub.slice(30, 38).toString()).toBe('mimetype');
  expect(epub.slice(38, 58).toString()).toBe('application/epub+zip');
  // Check the remaining files are compressed
  expect(epub.readUInt32BE(58)).toBe(0x504b0304);
  expect(epub.readUInt16LE(66)).not.toBe(0);
}

it('generate EPUB from single HTML with pub manifest', async () => {
  const manifest: PublicationManifest = {
    '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
    conformsTo: 'yuno',
    readingProgression: 'rtl',
    resources: [
      { url: 'cover%20image%25.png', rel: 'cover' },
      { url: 'index.html', rel: 'pagelist' },
    ],
    author: {
      name: [
        { value: 'Kenji Miyazawa', language: 'en' },
        { value: '宮沢賢治', language: 'ja' },
      ],
    },
    creator: ['foo', 'bar'],
    editor: 'baz',
    artist: 'a',
    illustrator: 'i',
    colorist: 'x',
    penciler: 'q',
    inker: 'c',
    letterer: 'b',
    translator: 'l',
    readBy: 'o',
    publisher: { name: 'Publisher', type: 'Organization' },
    contributor: 'Contributor',
    copyrightHolder: 'Acme Corporation',
    copyrightYear: '2023',
    subject: 'Subject',
  };
  vol.fromJSON({
    '/work/input/publication.json': JSON.stringify(manifest),
    '/work/input/index.html': /* html */ `
      <html lang="ja-JP">
      <head>
        <title>Document</title>
        <link rel="stylesheet" type="text/css" href="assets/style.css">
        <link rel="publication" type="application/ld+json" href="publication.json">
      </head>
      <body>
        <div role="doc-toc">
          <h1>Table of Contents</h1>
          <ol>
            <li><a href="#intro">Intro</a></li>
            <li>
              <a href="#main">Main</a>
              <ol>
                <li><a href="#main-1">Main 1</a></li>
                <li><a href="#main-2">Main 2</a></li>
              </ol>
            </li>
          </ol>
        </div>
        <div role="doc-pagelist">
          <ol>
            <li><a href="#intro">Intro</a></li>
            <li><a href="#main">Main</a></li>
            <li><a href="#main-1">Main 1</a></li>
            <li><a href="#main-2">Main 2</a></li>
          </ol>
        </div>
        <math></math>
        <svg></svg>
        <script src="https://example.com/remote-script.js"></script>
      </body>
      </html>
    `,
    '/work/input/cover image%.png': '',
  });
  await exportEpub({
    webpubDir: '/work/input',
    entryHtmlFile: 'index.html',
    manifest: decodePublicationManifest(manifest),
    relManifestPath: 'publication.json',
    target: '/work/output.epub',
    epubVersion: '3.0',
  });

  expect(toTree(vol)).toMatchSnapshot('tree');
  const file = vol.toJSON();
  expect(file['/tmp/1/META-INF/container.xml']).toMatchSnapshot(
    'container.xml',
  );
  expect(
    file['/tmp/1/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  const entry = file['/tmp/1/EPUB/index.xhtml'];
  assert(entry);
  expect(
    await format(entry, {
      parser: 'html',
      htmlWhitespaceSensitivity: 'strict',
    }),
  ).toMatchSnapshot('index.xhtml');
});

it('generate EPUB from series of HTML files', async () => {
  const manifest: PublicationManifest = {
    '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
    conformsTo: 'yuno',
    readingOrder: [
      'src/index.html',
      'src/a/index.html',
      'src/b/c/d.html',
      'src/escape%20check%25.html',
    ],
  };
  vol.fromJSON({
    '/work/input/publication.json': JSON.stringify(manifest),
    '/work/input/src/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>My book</title>
      </head>
      <body>
        <a href="#foo">1</a>
        <a href="a">2</a>
        <a href="./b/c/d/">3</a>
        <a href="./././escape%20check%25.html">3</a>
      </body>
      </html>
    `,
    '/work/input/src/a/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>yuno</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/src/b/c/d.html': /* html */ `
      <html lang="en">
      <head>
        <title>yunocchi</title>
      </head>
      <body></body>
      </html>
    `,
    '/work/input/src/escape check%.html': /* html */ `
      <html lang="en">
      <head>
        <title>日本語</title>
      </head>
      <body></body>
      </html>
    `,
  });
  await exportEpub({
    webpubDir: '/work/input',
    manifest: decodePublicationManifest(manifest),
    relManifestPath: 'publication.json',
    target: '/work/output.epub',
    epubVersion: '3.0',
  });

  expect(toTree(vol)).toMatchSnapshot('tree');
  const file = vol.toJSON();
  expect(
    file['/tmp/1/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  const first = file['/tmp/1/EPUB/src/index.xhtml'];
  assert(first);
  expect(
    await format(first, {
      parser: 'html',
      htmlWhitespaceSensitivity: 'strict',
    }),
  ).toMatchSnapshot('src/index.xhtml');
});

it('generate EPUB from single Markdown input', async () => {
  vol.fromJSON({
    '/work/input/foo bar%.md': '# 日本語',
  });
  await runCommand(['build', 'input/foo bar%.md', '-o', 'output.epub'], {
    cwd: '/work',
  });
  expect(toTree(vol)).toMatchSnapshot('tree');

  const epub = vol.readFileSync('/work/output.epub') as Buffer;
  checkValidEpubZip(epub);
  const zipFiles = new AdmZip(epub).getEntries();
  expect(
    zipFiles.reduce((acc, z) => {
      acc[z.entryName] = z.getData().toString();
      return acc;
    }, {}),
  ).toEqual({
    mimetype: 'application/epub+zip',
    ...vol.toJSON('/tmp/2', undefined, true),
  });
  const file = vol.toJSON();
  expect(
    file['/tmp/2/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  expect(file['/tmp/2/EPUB/foo bar%.xhtml']).toMatchSnapshot('foo bar%.xhtml');
});

it('generate EPUB from vivliostyle.config.js', async () => {
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify({
      entry: '日本語.md',
      output: './output.epub',
      theme: './escape check%.css',
      toc: {
        title: 'もくじ',
        htmlPath: 'gen content%/index file%.html',
        sectionDepth: 2,
      },
      cover: {
        src: 'cover image%.png',
        htmlPath: 'gen content%/cover document%.html',
      },
    }),
    '/work/input/日本語.md': `
# 日本語
## Sec. 1.1
### Sec. 1.1.1
    `,
    '/work/input/cover image%.png': '',
    '/work/input/escape check%.css': '/* theme CSS */',
  });
  await runCommand(['build'], { cwd: '/work/input' });

  expect(toTree(vol)).toMatchSnapshot('tree');

  const epub = vol.readFileSync('/work/input/output.epub') as Buffer;
  checkValidEpubZip(epub);
  const zipFiles = new AdmZip(epub).getEntries();
  expect(
    zipFiles.reduce((acc, z) => {
      acc[z.entryName] = z.getData().toString();
      return acc;
    }, {}),
  ).toEqual({
    mimetype: 'application/epub+zip',
    ...vol.toJSON('/tmp/2', undefined, true),
  });
  const file = vol.toJSON();
  expect(
    file['/tmp/2/EPUB/content.opf']
      ?.replace(/<dc:identifier id="bookid">.+<\/dc:identifier>/g, '')
      .replace(/<meta property="dcterms:modified">.+<\/meta>/g, ''),
  ).toMatchSnapshot('content.opf');
  const coverDocument = file['/tmp/2/EPUB/gen content%/cover document%.xhtml'];
  assert(coverDocument);
  expect(
    await format(coverDocument, {
      parser: 'html',
      htmlWhitespaceSensitivity: 'strict',
    }),
  ).toMatchSnapshot('cover document%.xhtml');
  const indexFile = file['/tmp/2/EPUB/gen content%/index file%.xhtml'];
  assert(indexFile);
  expect(
    await format(indexFile, {
      parser: 'html',
      htmlWhitespaceSensitivity: 'strict',
    }),
  ).toMatchSnapshot('index file%.xhtml');
});

it('Do not insert nav element to HTML that have nav[epub:type]', async () => {
  vol.fromJSON({
    '/work/input/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document</title>
      </head>
      <body>
        <nav epub:type="lot"></nav>
      </body>
      </html>
    `,
  });
  await runCommand(['build', 'index.html', '-o', 'output.epub'], {
    cwd: '/work/input',
  });

  const file = vol.toJSON();
  const xhtml = file['/tmp/2/EPUB/index.xhtml'];
  expect(xhtml).toMatch(/epub:type="lot"/);
  expect(xhtml).not.toMatch(/epub:type="toc"/);
  expect(xhtml).not.toMatch(/epub:type="landmarks"/);
});
