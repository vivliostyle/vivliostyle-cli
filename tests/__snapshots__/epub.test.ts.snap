// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate EPUB from series of HTML files > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"en\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>My book</dc:title>
    <dc:language>en</dc:language>
    
  </metadata>
  <manifest>
    <item id=\\"srcindexxhtml\\" href=\\"src/index.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"></item>
    <item id=\\"srcaindexxhtml\\" href=\\"src/a/index.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"srcbcdxhtml\\" href=\\"src/b/c/d.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"srcescape-checkxhtml\\" href=\\"src/escape%20check%25.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
  </manifest>
  <spine>
    <itemref idref=\\"srcindexxhtml\\"></itemref>
    <itemref idref=\\"srcaindexxhtml\\"></itemref>
    <itemref idref=\\"srcbcdxhtml\\"></itemref>
    <itemref idref=\\"srcescape-checkxhtml\\"></itemref>
  </spine>
</package>
"
`;

exports[`generate EPUB from series of HTML files > src/index.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  lang=\\"en\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
  ><head>
    <title>My book</title>
  </head>
  <body>
    <a href=\\"index.xhtml#foo\\">1</a>
    <a href=\\"a\\">2</a>
    <a href=\\"b/c/d\\">3</a>
    <a href=\\"escape%20check%25.xhtml\\">3</a>

    <nav id=\\"toc\\" role=\\"doc-toc\\" epub:type=\\"toc\\" hidden=\\"\\"
      ><h2>Table of Contents</h2
      ><ol
        ><li><a href=\\"index.xhtml\\">My book</a></li
        ><li><a href=\\"a/index.xhtml\\">yuno</a></li
        ><li><a href=\\"b/c/d.xhtml\\">yunocchi</a></li
        ><li><a href=\\"escape%20check%25.xhtml\\">日本語</a></li></ol
      ></nav
    ><nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\"
      ><h2>Landmarks</h2
      ><ol
        ><li
          ><a epub:type=\\"toc\\" href=\\"index.xhtml#toc\\">Table of Contents</a></li
        ></ol
      ></nav
    ></body
  ></html
>
"
`;

exports[`generate EPUB from series of HTML files > tree 1`] = `
"/
├─ tmp/
│  └─ 1/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  └─ src/
│     │     ├─ a/
│     │     │  └─ index.xhtml
│     │     ├─ b/
│     │     │  └─ c/
│     │     │     └─ d.xhtml
│     │     ├─ escape check%.xhtml
│     │     └─ index.xhtml
│     └─ META-INF/
│        └─ container.xml
└─ work/
   ├─ input/
   │  ├─ publication.json
   │  └─ src/
   │     ├─ a/
   │     │  └─ index.html
   │     ├─ b/
   │     │  └─ c/
   │     │     └─ d.html
   │     ├─ escape check%.html
   │     └─ index.html
   └─ output.epub"
`;

exports[`generate EPUB from single HTML with pub manifest > container.xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<container version=\\"1.0\\" xmlns=\\"urn:oasis:names:tc:opendocument:xmlns:container\\">
  <rootfiles>
    <rootfile full-path=\\"EPUB/content.opf\\" media-type=\\"application/oebps-package+xml\\" />
  </rootfiles>
</container>"
`;

exports[`generate EPUB from single HTML with pub manifest > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"ja-JP\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>Document</dc:title>
    <dc:language>ja-JP</dc:language>
    <dc:creator id=\\"author-1\\">宮沢賢治</dc:creator>
    <dc:creator id=\\"creator-1\\">foo</dc:creator>
    <dc:creator id=\\"creator-2\\">bar</dc:creator>
    <dc:creator id=\\"editor-1\\">baz</dc:creator>
    <dc:creator id=\\"artist-1\\">a</dc:creator>
    <dc:creator id=\\"illustrator-1\\">i</dc:creator>
    <dc:creator id=\\"colorist-1\\">x</dc:creator>
    <dc:creator id=\\"penciler-1\\">q</dc:creator>
    <dc:creator id=\\"inker-1\\">c</dc:creator>
    <dc:creator id=\\"letterer-1\\">b</dc:creator>
    <dc:creator id=\\"translator-1\\">l</dc:creator>
    <dc:creator id=\\"readby-1\\">o</dc:creator>
    <dc:publisher id=\\"publisher-1\\">Publisher</dc:publisher>
    <dc:contributor id=\\"contributor-1\\">Contributor</dc:contributor>
    <dc:rights>© 2023 Acme Corporation</dc:rights>
    <dc:subject>Subject</dc:subject>
    
    <meta name=\\"cover\\" content=\\"cover-imagepng\\"></meta>
  </metadata>
  <manifest>
    <item id=\\"cover-imagepng\\" href=\\"cover%20image%25.png\\" media-type=\\"image/png\\" properties=\\"cover-image\\"></item>
    <item id=\\"indexxhtml\\" href=\\"index.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav mathml remote-resources scripted svg\\"></item>
  </manifest>
  <spine page-progression-direction=\\"rtl\\">
    <itemref idref=\\"indexxhtml\\"></itemref>
  </spine>
</package>
"
`;

exports[`generate EPUB from single HTML with pub manifest > index.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  lang=\\"ja-JP\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
  ><head>
    <title>Document</title>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"assets/style.css\\" />
  </head>
  <body>
    <nav role=\\"doc-toc\\" id=\\"toc\\" epub:type=\\"toc\\">
      <h1>Table of Contents</h1>
      <ol>
        <li><a href=\\"index.xhtml#intro\\">Intro</a></li>
        <li>
          <a href=\\"index.xhtml#main\\">Main</a>
          <ol>
            <li><a href=\\"index.xhtml#main-1\\">Main 1</a></li>
            <li><a href=\\"index.xhtml#main-2\\">Main 2</a></li>
          </ol>
        </li>
      </ol>
    </nav>
    <nav role=\\"doc-pagelist\\" id=\\"page-list\\" epub:type=\\"page-list\\">
      <ol>
        <li><a href=\\"index.xhtml#intro\\">Intro</a></li>
        <li><a href=\\"index.xhtml#main\\">Main</a></li>
        <li><a href=\\"index.xhtml#main-1\\">Main 1</a></li>
        <li><a href=\\"index.xhtml#main-2\\">Main 2</a></li>
      </ol>
    </nav>
    <math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" />
    <svg xmlns=\\"http://www.w3.org/2000/svg\\" />
    <script src=\\"https://example.com/remote-script.js\\"></script>

    <nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\"
      ><h2>Landmarks</h2
      ><ol
        ><li
          ><a epub:type=\\"toc\\" href=\\"index.xhtml#toc\\">Table of Contents</a></li
        ></ol
      ></nav
    ></body
  ></html
>
"
`;

exports[`generate EPUB from single HTML with pub manifest > tree 1`] = `
"/
├─ tmp/
│  └─ 1/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  ├─ cover image%.png
│     │  └─ index.xhtml
│     └─ META-INF/
│        └─ container.xml
└─ work/
   ├─ input/
   │  ├─ cover image%.png
   │  ├─ index.html
   │  └─ publication.json
   └─ output.epub"
`;

exports[`generate EPUB from single Markdown input > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"en\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>日本語</dc:title>
    <dc:language>en</dc:language>
    
  </metadata>
  <manifest>
    <item id=\\"foo-barxhtml\\" href=\\"foo%20bar%25.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"></item>
  </manifest>
  <spine>
    <itemref idref=\\"foo-barxhtml\\"></itemref>
  </spine>
</package>
"
`;

exports[`generate EPUB from single Markdown input > foo bar%.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<!DOCTYPE html><html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:epub=\\"http://www.idpf.org/2007/ops\\"><head>
    <meta charset=\\"utf-8\\" />
    <title>日本語</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
  </head>
  <body>
    <section class=\\"level1\\" aria-labelledby=\\"日本語\\">
      <h1 id=\\"日本語\\">日本語</h1>
    </section>
  

<nav id=\\"toc\\" role=\\"doc-toc\\" epub:type=\\"toc\\" hidden=\\"\\"><h2>Table of Contents</h2><ol><li><a href=\\"foo%20bar%25.xhtml\\">日本語</a></li></ol></nav><nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\"><h2>Landmarks</h2><ol><li><a epub:type=\\"toc\\" href=\\"foo%20bar%25.xhtml#toc\\">Table of Contents</a></li></ol></nav></body></html>"
`;

exports[`generate EPUB from single Markdown input > tree 1`] = `
"/
├─ tmp/
│  ├─ 1/
│  │  ├─ foo bar%.html
│  │  └─ publication.json
│  └─ 2/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  └─ foo bar%.xhtml
│     └─ META-INF/
│        └─ container.xml
└─ work/
   ├─ input/
   │  └─ foo bar%.md
   └─ output.epub"
`;

exports[`generate EPUB from vivliostyle.config.js > content.opf 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"bookid\\" xml:lang=\\"en\\">
  <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
    
    <dc:title>日本語</dc:title>
    <dc:language>en</dc:language>
    
    <meta name=\\"cover\\" content=\\"cover-imagepng\\"></meta>
  </metadata>
  <manifest>
    <item id=\\"gen-contentcover-documentxhtml\\" href=\\"gen%20content%25/cover%20document%25.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"gen-contentindex-filexhtml\\" href=\\"gen%20content%25/index%20file%25.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"></item>
    <item id=\\"日本語xhtml\\" href=\\"%E6%97%A5%E6%9C%AC%E8%AA%9E.xhtml\\" media-type=\\"application/xhtml+xml\\"></item>
    <item id=\\"cover-imagepng\\" href=\\"cover%20image%25.png\\" media-type=\\"image/png\\" properties=\\"cover-image\\"></item>
    <item id=\\"escape-checkcss\\" href=\\"escape%20check%25.css\\" media-type=\\"text/css\\"></item>
  </manifest>
  <spine>
    <itemref idref=\\"gen-contentcover-documentxhtml\\"></itemref>
    <itemref idref=\\"gen-contentindex-filexhtml\\"></itemref>
    <itemref idref=\\"日本語xhtml\\"></itemref>
  </spine>
</package>
"
`;

exports[`generate EPUB from vivliostyle.config.js > cover document%.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
  ><head
    ><meta charset=\\"utf-8\\" /><title></title
    ><style data-vv-style=\\"\\">
      body {
        margin: 0;
      }
      [role=\\"doc-cover\\"] {
        display: block;
        width: 100vw;
        height: 100vh;
        object-fit: contain;
      }
      @page {
        margin: 0;
      }
    </style></head
  ><body
    ><section role=\\"region\\" aria-label=\\"Cover\\"
      ><img
        role=\\"doc-cover\\"
        src=\\"../cover%20image%25.png\\"
        alt=\\"Cover image\\" /></section></body
></html>
"
`;

exports[`generate EPUB from vivliostyle.config.js > index file%.xhtml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<html
  xmlns=\\"http://www.w3.org/1999/xhtml\\"
  xmlns:epub=\\"http://www.idpf.org/2007/ops\\"
  ><head
    ><meta charset=\\"utf-8\\" /><title>日本語</title
    ><link
      rel=\\"stylesheet\\"
      type=\\"text/css\\"
      href=\\"../escape%20check%25.css\\" /></head
  ><body
    ><h1>日本語</h1
    ><nav id=\\"toc\\" role=\\"doc-toc\\" epub:type=\\"toc\\"
      ><h2>もくじ</h2
      ><ol
        ><li data-section-level=\\"1\\"
          ><a
            href=\\"../%E6%97%A5%E6%9C%AC%E8%AA%9E.xhtml#%E6%97%A5%E6%9C%AC%E8%AA%9E\\"
            >日本語</a
          ><ol
            ><li data-section-level=\\"2\\"
              ><a href=\\"../%E6%97%A5%E6%9C%AC%E8%AA%9E.xhtml#sec-11\\"
                >Sec. 1.1</a
              ></li
            ></ol
          ></li
        ></ol
      ></nav
    ><nav epub:type=\\"landmarks\\" id=\\"landmarks\\" hidden=\\"\\"
      ><h2>Landmarks</h2
      ><ol
        ><li
          ><a epub:type=\\"toc\\" href=\\"index%20file%25.xhtml#toc\\"
            >Table of Contents</a
          ></li
        ><li
          ><a epub:type=\\"cover\\" href=\\"cover%20document%25.xhtml\\"
            >Cover Page</a
          ></li
        ></ol
      ></nav
    ></body
  ></html
>
"
`;

exports[`generate EPUB from vivliostyle.config.js > tree 1`] = `
"/
├─ tmp/
│  ├─ 1/
│  │  ├─ cover image%.png
│  │  ├─ escape check%.css
│  │  ├─ gen content%/
│  │  │  ├─ cover document%.html
│  │  │  └─ index file%.html
│  │  ├─ publication.json
│  │  └─ 日本語.html
│  └─ 2/
│     ├─ EPUB/
│     │  ├─ content.opf
│     │  ├─ cover image%.png
│     │  ├─ escape check%.css
│     │  ├─ gen content%/
│     │  │  ├─ cover document%.xhtml
│     │  │  └─ index file%.xhtml
│     │  └─ 日本語.xhtml
│     └─ META-INF/
│        └─ container.xml
└─ work/
   └─ input/
      ├─ .vivliostyle/
      │  ├─ cover image%.png
      │  ├─ escape check%.css
      │  ├─ gen content%/
      │  │  ├─ cover document%.html
      │  │  └─ index file%.html
      │  ├─ publication.json
      │  └─ 日本語.html
      ├─ cover image%.png
      ├─ escape check%.css
      ├─ output.epub
      ├─ vivliostyle.config.json
      └─ 日本語.md"
`;
